File: src/main/webapp/src/app/models/composition.ts
import { LeadSheet } from "./lead-sheet";
import { CompositionVideoFile } from "./composition-video-file";
import { CompositionMidiFile } from "./composition-midi-file";
import { CompositionAudioFile } from "./composition-audio-file";
import { CompositionFile } from "./composition-file";
import { ActionTriggerComposition } from "./action-trigger-composition";

export class Composition {
  name: string;
  durationMillis: number;
  fileList: CompositionFile[] = [];
  notes: string;
  autoStartNextComposition: boolean = false;
  leadSheetList: LeadSheet[] = [];
  loop: boolean;
  timecodeStartMillis: number = 0;
  // Only used while editing, so typing a timecode is not fought by the formatter
  timecodeStartText: string;
  midiNumber: number;
  showControlCue: string;
  audioVolume: number = 1;
  actionTriggerList: ActionTriggerComposition[] = [];

  constructor(data?: any) {
    if (!data) {
      return;
    }

    this.name = data.name;
    this.durationMillis = data.durationMillis;
    this.notes = data.notes;
    this.autoStartNextComposition = data.autoStartNextComposition;

    this.fileList = this.parseFileList(data);

    if (data.leadSheetList) {
      for (let leadSheet of data.leadSheetList) {
        this.leadSheetList.push(new LeadSheet(leadSheet));
      }
    }

    this.loop = data.loop;
    this.timecodeStartMillis = data.timecodeStartMillis || 0;
    this.midiNumber = data.midiNumber;
    this.showControlCue = data.showControlCue;
    this.audioVolume = data.audioVolume;

    if (data.actionTriggerList) {
      this.actionTriggerList = [];

      for (let actionTrigger of data.actionTriggerList) {
        this.actionTriggerList.push(
          new ActionTriggerComposition(actionTrigger)
        );
      }
    }
  }

  public static getFileObjectByType(data: any) {
    if (data.midiFile) {
      let midiFile = new CompositionMidiFile(data.midiFile);
      return midiFile;
    } else if (data.audioFile) {
      let audioFile = new CompositionAudioFile(data.audioFile);
      return audioFile;
    } else if (data.videoFile) {
      let videoFile = new CompositionVideoFile(data.videoFile);
      return videoFile;
    }
  }

  private parseFileList(data: any): CompositionFile[] {
    let fileList: CompositionFile[] = [];

    if (data.fileList) {
      for (let file of data.fileList) {
        fileList.push(Composition.getFileObjectByType(file));
      }
    }

    return fileList;
  }

  toJSON() {
    return {
      ...this,
      // Exclude unwanted properties
      autoStartNextComposition: undefined,
      timecodeStartText: undefined,
    };
  }
}

File: src/main/java/com/ascargon/rocketshow/composition/Composition.java
package com.ascargon.rocketshow.composition;

import com.ascargon.rocketshow.audio.AudioCompositionFile;
import com.ascargon.rocketshow.midi.MidiCompositionFile;
import com.ascargon.rocketshow.video.VideoCompositionFile;
import com.fasterxml.jackson.annotation.JsonProperty;

import jakarta.xml.bind.annotation.XmlElement;
import jakarta.xml.bind.annotation.XmlElementWrapper;
import jakarta.xml.bind.annotation.XmlElements;
import jakarta.xml.bind.annotation.XmlRootElement;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;

@XmlRootElement
@Getter
@Setter
public class Composition {

    private String name;
    private String notes;
    private long durationMillis;
    private boolean loop = false;

    // The absolute MIDI timecode position (in milliseconds) at which this composition starts, used
    // when following an incoming MIDI timecode with the per-composition mapping.
    private long timecodeStartMillis = 0;

    // The number that selects this composition over MIDI (program change or song select), counted
    // from 1. Null if this composition cannot be selected over MIDI.
    private Integer midiNumber;

    // The MIDI Show Control cue number that selects this composition, e.g. "12.5". Empty to use the
    // MIDI number instead.
    private String showControlCue;
    private List<CompositionFile> compositionFileList = new ArrayList<>();
    private List<LeadSheet> leadSheetList = new ArrayList<>();
    private float audioVolume = 1;
    private List<ActionTriggerComposition> actionTriggerList = new ArrayList<>();

    @XmlElementWrapper(name = "fileList")
    @XmlElements({@XmlElement(type = MidiCompositionFile.class, name = "midiFile"),
            @XmlElement(type = VideoCompositionFile.class, name = "videoFile"),
            @XmlElement(type = AudioCompositionFile.class, name = "audioFile")})
    @JsonProperty("fileList")
    @SuppressWarnings("WeakerAccess")
    public List<CompositionFile> getCompositionFileList() {
        return compositionFileList;
    }

}
